### PR TITLE
feat(vocabulary): vocab support for map in vocab auto generation

### LINKED ISSUE #69: Supporting Map type declaration, key and value for vocabulary codegen
## Steps to Reproduce

1.
2.
3.
4.

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/concerto)
- [ ] [GitHub Issues](https://github.com/accordproject/concerto/issues)

## Your Environment

* Version used:
* Environment name and version (e.g. Chrome 39, node.js 5.4):
* Operating System and version (desktop or mobile):
* Link to your project:

### PR DESCRIPTION
**Pull request by sanketshevkar**
# Closes #69 

### Changes

- added visitor methods for map-declaration. map-key, map-value

### Related Issues
- Issue #69 

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [x] Extend the documentation, if necessary
- [x] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: lib/codegen/fromcto/vocabulary/vocabularyvisitor.js [hunk 1]
```diff
@@ -141,6 +145,30 @@ class VocabularyVisitor {
...
+     * @param {MapDeclaration} mapDeclaration - the object being visited
+     * @param {Object} parameters  - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitMapDeclaration(mapDeclaration, parameters) {
+        const mapDeclarationName = mapDeclaration.getName();
+        const mapDeclarationVocabulary = VocabularyManager.englishMissingTermGenerator(null, null, mapDeclarationName);
+
+        parameters.fileWriter.writeLine(0, `  - ${mapDeclarationName}: ${mapDeclarationVocabulary}`);
```
Comment (ekarademir) [inline lib/codegen/fromcto/vocabulary/vocabularyvisitor.js:159]: Are the whitespaces before these lines standard? i.e., if this is adding to an existing vocab yaml file, will it be at the same indentation level as the one in the yaml file?

File: lib/codegen/fromcto/vocabulary/vocabularyvisitor.js [hunk 2]
```diff
@@ -54,6 +54,10 @@ class VocabularyVisitor {
             return this.visitProperty(thing, parameters);
         } else if (thing.isEnumValue?.()) {
             return this.visitProperty(thing, parameters);
+        } else if (thing.isKey?.()) {
+            return this.visitMapKey(thing, parameters);
+        } else if (thing.isValue?.()) {
```
Comment (ekarademir) [inline lib/codegen/fromcto/vocabulary/vocabularyvisitor.js:59]: I guess, it's a bit late to do anything about it now, but `isValue` looks a lot more generic than it should be. We could have made it like `isMapValue`, just as we did for declarations. Just a remark, nothing actionable.

Comment (ekarademir) [reply lib/codegen/fromcto/vocabulary/vocabularyvisitor.js:59]: Maybe to future-proof this, if possible, we can check if the parent of the `thing` is a map, os something similar to scope this to map declarations only. The same goes for the keys.

File: lib/codegen/fromcto/vocabulary/vocabularyvisitor.js (hunk 1 above)
Comment (sanketshevkar) [reply lib/codegen/fromcto/vocabulary/vocabularyvisitor.js:159]: I don't think we support passing an existing YAML file to this. The purpose of `VocabularyVistitor` is just to bootstrap a new `.voc` file.

File: lib/codegen/fromcto/vocabulary/vocabularyvisitor.js (hunk 2 above)
Comment (sanketshevkar) [reply lib/codegen/fromcto/vocabulary/vocabularyvisitor.js:59]: For now I believe keys and values are only scoped to Maps as of now.

Comment (sanketshevkar) [reply lib/codegen/fromcto/vocabulary/vocabularyvisitor.js:59]: If you are suggesting for readability and interpretation, that's why kept the visitor method names as `visitMapKey` `visitMapValue`.
I agree with you the method name is very generic.

**Comment by Jonathan-Casey** [comment]
@sanketshevkar as an aside, you don't necessarily need to use recursion to visit the Map Key & Value, you could do it all within the VisitMapDeclaration. LGTM, nice work!

**Review by Jonathan-Casey — APPROVED** [review]

**Comment by sanketshevkar** [comment]
> @sanketshevkar as an aside, you don't necessarily need to use recursion to visit the Map Key & Value, you could do it all within the VisitMapDeclaration and keep things a bit more compact. LGTM, nice work!

Thanks @Jonathan-Casey! Agreed with your suggestion. But I think we will on have two recursive calls inside a map declaration level. Shouldn't cause a big issue I believe.
Shall I go forward with merging the code?

**Review by ekarademir — APPROVED** [review]

**Pull request merged**